sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SideInputLoadTest.java

```diff
@@ -0,0 +1,243 @@
+      PCollection<KV<byte[], byte[]>> input,
+      SideInputMaterializationType sideInputType,
+      Optional<SyntheticStep> syntheticStep) {
+    switch (sideInputType) {
+      case ITERABLE:
+        performTestWithIterable(input, syntheticStep);
+        break;
+      case HASHMAP:
+        performTestWithHashMap(input, syntheticStep);
+        break;
+      case LIST:
+        performTestWithList(input, syntheticStep);
+        break;
+    }
+  }
+
+  private void performTestWithList(
+      PCollection<KV<byte[], byte[]>> input, Optional<SyntheticStep> syntheticStep) {
+    applyStepIfPresent(input, "Synthetic step", syntheticStep);
+    PCollectionView<List<KV<byte[], byte[]>>> sideInput = input.apply(View.asList());
+    input
+        .apply(ParDo.of(new SideInputTestWithList(sideInput)).withSideInputs(sideInput))
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
+  }
+
+  private void performTestWithHashMap(
+      PCollection<KV<byte[], byte[]>> input, Optional<SyntheticStep> syntheticStep) {
+    applyStepIfPresent(input, "Synthetic step", syntheticStep);
+    PCollectionView<Map<byte[], byte[]>> sideInput = input.apply(View.asMap());
+    input
+        .apply(ParDo.of(new SideInputTestWithHashMap(sideInput)).withSideInputs(sideInput))
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
+  }
+
+  private void performTestWithIterable(
+      PCollection<KV<byte[], byte[]>> input, Optional<SyntheticStep> syntheticStep) {
+    applyStepIfPresent(input, "Synthetic step", syntheticStep);
+    PCollectionView<Iterable<KV<byte[], byte[]>>> sideInput;
+    if (options.getWindowCount() != 0) {
+      long windowDurationMilis = sourceOptions.numRecords / options.getWindowCount();
+      sideInput =
+          input
+              .apply(Window.into(FixedWindows.of(Duration.millis(windowDurationMilis))))
+              .apply(View.asIterable());
+    } else {
+      sideInput = input.apply(View.asIterable());
+    }
+
+    input
+        .apply(ParDo.of(new SideInputTestWithIterable(sideInput)).withSideInputs(sideInput))
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
+  }
+
+  private static class AddTimestamps extends DoFn<KV<byte[], byte[]>, KV<byte[], byte[]>> {
+    private static Instant timestamp = TIME;
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      timestamp = TIME.plus(1L);
+      c.outputWithTimestamp(c.element(), timestamp);
+    }
+  }
+
+  private static class SideInputTestWithList extends DoFn<KV<byte[], byte[]>, KV<byte[], byte[]>> {
+
+    private final PCollectionView<List<KV<byte[], byte[]>>> sideInput;
+
+    public SideInputTestWithList(PCollectionView<List<KV<byte[], byte[]>>> sideInput) {
+      this.sideInput = sideInput;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      List<KV<byte[], byte[]>> si = c.sideInput(sideInput);
+
+      for (KV<byte[], byte[]> sideInputElement : si) {
```

Comment:
It is worth documenting that the whole point of `View.asList` is to force it all to be read into memory so it is fast to iterate multiple times or access a given index. You could test these, too. If you don't need either of those, then `View.asIterable` will access lazily so it will often perform better and scale larger.